src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -676,38 +676,191 @@ private static IEnumerable<TSource> TakeWhileIterator<TSource>(IEnumerable<TSour
+            internal DeadEnumeratorWrapper(IEnumerator<TSource> corpse)
+                : this(corpse, null)
+            {
+            }
+
+            internal DeadEnumeratorWrapper(IEnumerator<TSource> corpse, ExceptionDispatchInfo exception)
+            {
+                _corpse = corpse;
+                _exception = exception;
+            }
+
+            public TSource Current
+            {
+                get { return default(TSource); }
+            }
+
+            object IEnumerator.Current
+            {
+                get { return Current; }
+            }
+
+            public bool MoveNext()
+            {
+                if (_exception != null)
+                {
+                    ExceptionDispatchInfo ex = _exception;
+                    _exception = null;
+                    ex.Throw();
+                }
+                return false;
+            }
+
+            public void Dispose()
+            {
+                _corpse.Dispose();
+            }
+
+            public void Reset()
+            {
+                throw new NotSupportedException();
+            }
+        }
+
         public static IEnumerable<TSource> Skip<TSource>(this IEnumerable<TSource> source, int count)
         {
             if (source == null) throw Error.ArgumentNull("source");
-            return SkipIterator<TSource>(source, count);
+            return new SkipIterator<TSource>(source, count);
         }
 
-        private static IEnumerable<TSource> SkipIterator<TSource>(IEnumerable<TSource> source, int count)
+        private sealed class SkipIterator<TSource> : IEnumerable<TSource>
         {
-            using (IEnumerator<TSource> e = source.GetEnumerator())
+            private readonly IEnumerable<TSource> _source;
+            private readonly int _count;
+
+            public SkipIterator(IEnumerable<TSource> source, int count)
             {
-                while (count > 0 && e.MoveNext()) count--;
-                if (count <= 0)
+                _source = source;
+                _count = count > 0 ? count : 0;
+            }
+
+            public IEnumerator<TSource> GetEnumerator()
+            {
+                IEnumerator<TSource> sourceEnumerator = _source.GetEnumerator();
+                try
                 {
-                    while (e.MoveNext()) yield return e.Current;
+                    for (int count = _count; count != 0; --count)
+                    {
+                        // It should be harmless to have the recipient call MoveNext again.
+                        // but we guard against strange behaviour from shortcuts in the enumerator
+                        // It should also be harmless to dispose the enumerator here if we've
+                        // exhausted it, but we guard against strange behaviour in either the
+                        // enumerator or the caller.
+                        if (!sourceEnumerator.MoveNext()) return new DeadEnumeratorWrapper<TSource>(sourceEnumerator);
```

Comment:
Hmm. Calling `GetEnumerator()` and not calling `MoveNext()` even once almost immediately seems pretty obscure, even to someone who has a strange love for playing with enumerators like myself, but certainly not impossible.
The only way around that is to wrap it entirely. While that proved to be a win with `SkipWhile` I wouldn't expect it to be as big a win, and perhaps not worth the effort.